### ISSUE TITLE
Main OS would kill chipdb.py script

### ISSUE THREAD

**Issue by kxynos** [YosysHQ/nextpnr#445]
On my Linux system I had issues completing the compilation since the system would kill nextpnr/ice40/chipdb.py before it completed the task. 

Output of the issue : 
```
$ make -j$(nproc)
Scanning dependencies of target bbasm
[  1%] Building CXX object CMakeFiles/bbasm.dir/bba/main.cc.o
[  2%] Linking CXX executable bbasm
[  2%] Built target bbasm
[  3%] Generating ice40/chipdbs/chipdb-384.bba
[  4%] Generating ice40/chipdbs/chipdb-1k.bba
[  5%] Generating ice40/chipdbs/chipdb-5k.bba
[  6%] Generating ice40/chipdbs/chipdb-u4k.bba
[  7%] Generating ice40/chipdbs/chipdb-8k.bba
Killed
make[2]: *** [CMakeFiles/ice40_chipdb.dir/build.make:131: ice40/chipdbs/chipdb-8k.bba] Error 137
make[1]: *** [CMakeFiles/Makefile2:165: CMakeFiles/ice40_chipdb.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
The problem is that the Python script output is piped to a file. This creates a number of issues. 

I quick fix is to redirect STDOUT to a file instead. I have implemented the required updates and tested them.

**Comment by whitequark** [comment]
> This creates a number of issues.

Could you detail which issues? Your fix seems reasonable to me but I don't understand why it's needed.

**Comment by smunaut** [comment]
This seems more like a random running out of memory and the patch either just steers the OOM to some other process or some memory got freed in the mean time or something that's not really the root cause of the issue.

**Comment by kxynos** [comment]
> > This creates a number of issues.
> 
> Could you detail which issues? Your fix seems reasonable to me but I don't understand why it's needed.

I had an error in dmesg of the process running out of memory and it being killed by the OS. The system really suffered as nextpnr/ice40/chipdbs/chipdb-8k.bba was assembled. I tried it a few times and it got killed every time. With the modification I had no issues.

The speed up of saving to a file seems marginal on tests on run again just now. I did not run enough tests to say if the speedup is significant overall.

You are correct, something did go wrong. I think piping such a large file, with the processing that occurs, is not ideal. Hence the PR. 

Ideally the Python script should be writing the output to a file and not printing to the console. I used the if statement to maintain the ability to still print to the console when needed. So it is backwards compatible.

**Comment by whitequark** [comment]
> This seems more like a random running out of memory and the patch either just steers the OOM to some other process or some memory got freed in the mean time or something that's not really the root cause of the issue.

I believe the patch genuinely reduces memory pressure by avoiding buffering of the output on Python side of things, but I didn't verify that this is actually what's happening.

**Comment by smunaut** [comment]
It will buffer a bit more, (because in addition to the python buffer you'd have the pipe), but that should be negligible. If the write to the pipe isn't fast, it will apply back pressure on the writing app (python here).

For instance an infinite generator :

```python
#!/usr/bin/env python3
import sys
d = 1024 * ' ' + '\n';
while True:
	sys.stdout.write(d)
```

Then pipe it to `less` which won't read anything :

```
./gen.py | less
```

And the python is blocked very quickly, there is almost no output buffered. (I mean chipdb is in the order of Gbytes of mem usage, the pipe buffers should be in the 100's of Kbytes ...)

(note that I'm not saying writing to file directly isn't better, I'd just like to understand what's going on ...)

**Comment by whitequark** [comment]
> (note that I'm not saying writing to file directly isn't better, I'd just like to understand what's going on ...)

Same here. So my guess was incorrect. I'm not sure what's happening here then.

**Comment by kxynos** [comment]
Not sure if this will help. 

I did a simple test to track CPU and memory usage of the two commands, using psrecord. It might be that my system was low on memory and outputting it to a file didn't push it to the limits compared to piping to a file. 

It is clear that there is a higher memory usage for the one using file pipe. 

Time test shell script:
```
/usr/bin/python3 nextpnr/ice40/chipdb.py -p nextpnr/ice40/constids.inc -g nextpnr/ice40/gfx.h --fast /usr/local/share/icebox/timings_hx8k.txt --slow /usr/local/share/icebox/timings_lp8k.txt /usr/local/share/icebox/chipdb-8k.txt -o nextpnr/ice40/chipdbs/chipdb-8k.bba.new & psrecord $(pgrep python3) --interval 1 --plot 0x.output.png --log 0x.output.log

/usr/bin/python3 nextpnr/ice40/chipdb.py -p nextpnr/ice40/constids.inc -g nextpnr/ice40/gfx.h --fast /usr/local/share/icebox/timings_hx8k.txt --slow /usr/local/share/icebox/timings_lp8k.txt /usr/local/share/icebox/chipdb-8k.txt >> nextpnr/ice40/chipdbs/chipdb-8k.bba.new & psrecord $(pgrep python3) --interval 1 --plot 0x.pipe.png --log 0x.pipe.log 
```

File pipe timings(0x.pipe.log) snippet:
```
# Elapsed time   CPU (%)     Real (MB)   Virtual (MB)
       0.003        0.000       29.715       35.703
       1.006       98.700       93.113       99.078
       2.009       98.700      163.598      169.578
       3.014       99.500      215.340      221.328
       4.015       98.900      322.320      328.328
       5.018       97.700      410.457      416.410
       6.019       99.800      507.621      513.375
       7.023       98.700      593.625      599.504
       8.024       97.900      701.844      707.895
       9.026       98.800      746.777      752.688
      10.027       97.900      876.551      882.742
...
```

File output timings(0x.output.log) snippet:
```
# Elapsed time   CPU (%)     Real (MB)   Virtual (MB)
       0.002        0.000       26.074       31.953
       1.004       98.800       91.449       97.328
       2.006       97.800      143.469      149.328
       3.007       97.900      215.453      221.328
       4.009       97.700      306.219      312.078
       5.011       99.900      393.758      399.660
       6.013       98.800      498.051      503.977
       7.015       98.800      571.129      576.973
       8.017       97.800      679.824      685.762
       9.018       97.900      746.945      752.691
      10.025       98.400      856.250      901.434
...
```

Full file listings:

File pipe timings(0x.pipe.log):
```
# Elapsed time   CPU (%)     Real (MB)   Virtual (MB)
       0.000        0.000       29.223       35.203
       1.002       97.900       93.637       99.578
       2.004       99.800      164.836      170.828
       3.005       97.800      215.340      221.328
       4.007       98.800      287.336      293.328
       5.009       98.800      352.859      358.887
       6.012       98.700      470.727      476.773
       7.015       98.700      541.191      547.258
       8.017       97.800      649.996      655.785
       9.019       97.800      746.898      752.688
      10.021       98.800      778.672      784.707
      11.023       96.800      909.680      915.770
      12.025       99.800      999.977     1006.023
      13.027       97.800     1019.293     1025.273
      14.034       99.300     1093.285     1099.578
      15.037       98.700     1145.406     1151.578
      16.040       98.700     1196.941     1203.078
      17.042       99.800     1238.691     1244.828
      18.043       98.800     1299.676     1305.828
      19.045       98.900     1339.707     1345.828
      20.046       98.800     1379.910     1386.078
      21.047       98.900     1420.988     1427.078
      22.049       99.800     1500.414     1506.582
      23.051       97.800     1541.234     1547.582
      24.058       99.300     1581.453     1587.582
      25.062       98.700     1622.180     1628.332
      26.068       99.400     1661.672     1667.832
      27.076       99.200     1702.664     1708.832
      28.078       98.800     1742.930     1749.082
      29.079       98.900     1783.234     1789.582
      30.081       98.800     1822.672     1828.832
      31.085       98.600     1900.250     1906.332
      32.087       99.800     1940.703     1946.832
      33.087       97.900     1980.453     1986.832
      34.090       99.800     2019.422     2025.582
      35.091       95.800     2051.164     2057.332
      36.095       98.700     2076.164     2082.332
      37.096       98.900     2108.988     2115.082
      38.098       97.700     2108.988     2115.082
      39.099       98.900     2108.988     2115.082
      40.103       98.600     2108.988     2115.082
      41.105       97.900     2108.988     2115.082
      42.106       96.900     2108.988     2115.082
      43.109       97.700     2108.988     2115.082
      44.111       97.900     2108.988     2115.082
      45.112       97.900     2108.988     2115.082
      46.113       97.800     2108.988     2115.082
      47.115       97.800     2108.988     2115.082
      48.117       97.800     2108.988     2115.082
      49.118       97.900     2108.988     2115.082
      50.119       98.900     2108.988     2115.082
      51.121       97.900     2108.988     2115.082
      52.127       96.400     2108.988     2115.082
      53.128       97.800     2108.988     2115.082
      54.131       99.800     2108.988     2115.082
      55.133       96.800     2160.242     2166.332
      56.135       98.800     2160.242     2166.332
      57.141       94.400     2252.555     2258.832
      58.144       97.800     2252.555     2258.832
      59.146       98.800     2252.555     2258.832
      60.151       98.500     2252.812     2258.832
      61.152       97.900     2252.812     2258.832
      62.154       98.800     2252.812     2258.832
      63.155       98.900     1955.312     1961.324
[EOF]
```

File output timings(0x.output.log):
```
# Elapsed time   CPU (%)     Real (MB)   Virtual (MB)
       0.002        0.000       26.074       31.953
       1.004       98.800       91.449       97.328
       2.006       97.800      143.469      149.328
       3.007       97.900      215.453      221.328
       4.009       97.700      306.219      312.078
       5.011       99.900      393.758      399.660
       6.013       98.800      498.051      503.977
       7.015       98.800      571.129      576.973
       8.017       97.800      679.824      685.762
       9.018       97.900      746.945      752.691
      10.025       98.400      856.250      901.434
      11.026       98.800      957.543      963.590
      12.031       98.500     1000.074     1006.023
      13.033       98.800     1073.281     1079.328
      14.035       99.800     1129.805     1135.828
      15.038       98.700     1182.297     1188.328
      16.040       98.800     1223.562     1229.578
      17.041       98.900     1284.520     1290.578
      18.045       98.700     1325.812     1331.828
      19.047       98.800     1367.070     1373.078
      20.051       99.500     1404.117     1410.078
      21.052       98.900     1443.531     1449.578
      22.060       98.200     1523.309     1529.332
      23.061       98.800     1560.551     1566.582
      24.062       97.900     1598.613     1604.582
      25.065       99.700     1637.379     1643.332
      26.068       98.700     1677.152     1683.332
      27.070       99.800     1717.613     1723.582
      28.072       97.800     1757.309     1763.332
      29.076       98.600     1796.301     1802.332
      30.078       98.800     1835.887     1841.832
      31.078       98.900     1911.348     1917.332
      32.080       99.800     1946.879     1952.832
      33.081       97.900     1979.023     1985.082
      34.083       98.900     2013.332     2019.332
      35.084       98.900     2048.574     2054.582
      36.085       98.900     2083.809     2089.832
      37.086       97.900     2108.816     2115.082
      38.088       97.800     2108.816     2115.082
      39.090       97.800     2108.816     2115.082
      40.093       98.700     2109.133     2115.082
      41.094       96.800     2109.133     2115.082
      42.096       99.800     2109.133     2115.082
      43.099       98.700     2109.133     2115.082
      44.100       96.800     2109.133     2115.082
      45.108       99.300     2109.133     2115.082
      46.114       98.400     2109.133     2115.082
      47.117       99.700     2109.133     2115.082
      48.121       96.700     2109.133     2115.082
      49.123       98.700     2109.133     2115.082
      50.125       98.800     2109.133     2115.082
      51.127       95.900     2109.133     2115.082
      52.132       98.500     2109.133     2115.082
      53.134       96.800     2109.133     2115.082
      54.135       98.900     2109.133     2115.082
      55.139       96.600     2160.355     2166.332
      56.142       98.700     2160.355     2166.332
      57.144       96.800     2252.738     2258.832
      58.146       96.800     2252.738     2258.832
      59.150       97.600     2252.734     2258.832
      60.154       97.700     2252.734     2258.832
      61.157       98.700     2252.734     2258.832
      62.162       98.500     2252.734     2258.832
      63.167       98.500      248.168      254.070
[EOF]
```

**Comment by smunaut** [comment]
What do you mean by "It is clear that there is a higher memory usage for the one using file pipe."

Both show the exact same peak memory usage of 2252 Mb ...

**Comment by kxynos** [comment]
> What do you mean by "It is clear that there is a higher memory usage for the one using file pipe."
> 
> Both show the exact same peak memory usage of 2252 Mb ...
Correct. 
I want to highlight that the progression shows a higher usage from the start.

**Comment by daveshah1** [comment]
Given this doesn't actually seem to help much, I am closing the issue and PR.

**Comment by zzdts** [comment]
I meet this issue when i allocate 2GiB memory to my virtual machine, I solved when  allocate 5.8GiB memory to vm.
